sql/core/src/test/scala/org/apache/spark/sql/connector/MetadataColumnSuite.scala

```diff
@@ -232,4 +239,191 @@ class MetadataColumnSuite extends DatasourceV2SQLBase {
+      checkAnswer(dfQuery, expectedAnswer)
+    }
+  }
+
+  test("SPARK-41498: Metadata column is not propagated when children of Union " +
+    "have metadata output of different size") {
+    withTable(tbl) {
+      prepareTable()
+      withTempDir { dir =>
+        spark.range(start = 10, end = 20).selectExpr("bigint(id) as id", "string(id) as data")
+          .write.mode("overwrite").save(dir.getAbsolutePath)
+        val df1 = spark.table(tbl)
+        val df2 = spark.read.load(dir.getAbsolutePath)
+
+        // Make sure one df contains a metadata column and the other does not
+        assert(!df1.queryExecution.analyzed.metadataOutput.exists(_.name == "_metadata"))
+        assert(df2.queryExecution.analyzed.metadataOutput.exists(_.name == "_metadata"))
+
+        assert(df1.union(df2).queryExecution.analyzed.metadataOutput.isEmpty)
+      }
+    }
+  }
+}
+
+class MetadataTestTable extends Table with SupportsMetadataColumns {
+
+  override def name(): String = "fake"
+  override def schema(): StructType = StructType(Nil)
+  override def capabilities(): java.util.Set[TableCapability] =
+    java.util.EnumSet.of(TableCapability.BATCH_READ)
+  override def metadataColumns(): Array[MetadataColumn] =
+    Array(
+      new MetadataColumn {
+        override def name: String = "_metadata"
+        override def dataType: DataType = StructType(StructField("index", IntegerType) :: Nil)
+        override def comment: String = ""
+      }
+    )
+}
+
+class TypeMismatchTable extends MetadataTestTable {
+
+  override def metadataColumns(): Array[MetadataColumn] =
+    Array(
+      new MetadataColumn {
+        override def name: String = "_metadata"
+        override def dataType: DataType = StructType(StructField("index", StringType) :: Nil)
+        override def comment: String =
+          "Used to create a type mismatch with the metadata col in `MetadataTestTable`"
+      }
+    )
+}
+
+class AttrNameMismatchTable extends MetadataTestTable {
+  override def metadataColumns(): Array[MetadataColumn] =
+    Array(
+      new MetadataColumn {
+        override def name: String = "wrongName"
+        override def dataType: DataType = StructType(StructField("index", IntegerType) :: Nil)
+        override def comment: String =
+          "Used to create a name mismatch with the metadata col in `MetadataTestTable`"
+      })
+}
+
+class FieldNameMismatchTable extends MetadataTestTable {
+  override def metadataColumns(): Array[MetadataColumn] =
+    Array(
+      new MetadataColumn {
+        override def name: String = "_metadata"
+        override def dataType: DataType = StructType(StructField("wrongName", IntegerType) :: Nil)
+        override def comment: String =
+          "Used to create a name mismatch with the struct field in the metadata col of " +
+            "`MetadataTestTable`"
+      })
+}
+
+class MetadataTestCatalog extends TestV2SessionCatalogBase[MetadataTestTable] {
```

Comment:
It's better to extend `InMemoryCatalog`, as we don't need to override the session catalog to test this feature.